### PR TITLE
Refactor bookshelf view

### PR DESCRIPTION
**Pull request by kaansonmezoz**
## Summary of change
- Optimized imports.
- Refactored BookInShelfView
   - Seperated different logics from BookInShelfView to related classes.

## Related issue
#197 

## Pull request checklist

Please ensure you have done the following:

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request

- [x] Created a branch that has a descriptive name (what your branch is for a in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

If in doubt, get in touch with us via our [Gitter room](https://gitter.im/book-project-community/community)

### CODE REVIEW

File: src/main/java/com/karankumar/bookproject/backend/entity/Author.java [hunk 1]
```diff
@@ -15,11 +15,7 @@
 
 package com.karankumar.bookproject.backend.entity;
 
-import lombok.AccessLevel;
-import lombok.AllArgsConstructor;
-import lombok.Data;
-import lombok.EqualsAndHashCode;
-import lombok.NoArgsConstructor;
+import lombok.*;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/entity/Author.java:18]: Could you please revert this change? We are not using wildcard imports (see https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports)

File: src/main/java/com/karankumar/bookproject/backend/entity/Role.java [hunk 2]
```diff
@@ -15,20 +15,9 @@
...
 
-import lombok.AccessLevel;
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Data;
-import lombok.EqualsAndHashCode;
-import lombok.Getter;
-import lombok.NoArgsConstructor;
-import lombok.ToString;
+import lombok.*;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/entity/Role.java:18]: Another wildcard import to undo

File: src/main/java/com/karankumar/bookproject/backend/entity/Shelf.java [hunk 3]
```diff
@@ -1,10 +1,6 @@
 package com.karankumar.bookproject.backend.entity;
 
-import lombok.AccessLevel;
-import lombok.Getter;
-import lombok.NoArgsConstructor;
-import lombok.Setter;
-import lombok.ToString;
+import lombok.*;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/entity/Shelf.java:3]: Please undo this

File: src/main/java/com/karankumar/bookproject/backend/entity/User.java [hunk 4]
```diff
@@ -15,25 +15,9 @@
...
 
-import lombok.AccessLevel;
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Data;
-import lombok.EqualsAndHashCode;
-import lombok.Getter;
-import lombok.NoArgsConstructor;
-import lombok.ToString;
+import lombok.*;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/entity/User.java:18]: See previous wildcard import comment

File: src/main/java/com/karankumar/bookproject/backend/entity/User.java [hunk 5]
```diff
@@ -15,25 +15,9 @@
...
-import javax.persistence.Entity;
-import javax.persistence.FetchType;
-import javax.persistence.GeneratedValue;
-import javax.persistence.GenerationType;
-import javax.persistence.Id;
-import javax.persistence.JoinColumn;
-import javax.persistence.JoinTable;
-import javax.persistence.ManyToMany;
-import javax.persistence.SequenceGenerator;
+import javax.persistence.*;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/entity/User.java:20]: See previous wildcard import comment

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java [hunk 6]
```diff
@@ -16,35 +16,18 @@
...
-
-import org.springframework.stereotype.Service;
-
-import com.karankumar.bookproject.backend.entity.Author;
-import com.karankumar.bookproject.backend.entity.Book;
-import com.karankumar.bookproject.backend.entity.Genre;
-import com.karankumar.bookproject.backend.entity.PredefinedShelf;
-import com.karankumar.bookproject.backend.entity.RatingScale;
-import com.karankumar.bookproject.backend.entity.Tag;
+import com.karankumar.bookproject.backend.entity.*;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:19]: See previous wildcard import comment

File: src/main/java/com/karankumar/bookproject/ui/components/PaperToggle.java [hunk 7]
```diff
@@ -1,6 +1,8 @@
 package com.karankumar.bookproject.ui.components;
 
-import com.vaadin.flow.component.*;
+import com.vaadin.flow.component.Component;
+import com.vaadin.flow.component.ComponentEventListener;
+import com.vaadin.flow.component.Tag;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/components/PaperToggle.java:5]: See previous wildcard import comment

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    private boolean containsBookAuthor(Book book) {
+        return isBookAuthorNull() || book.getAuthor().toString().toLowerCase().contains(getBookAuthor());
+    }
+
+    private boolean containsBookTitle(Book book) {
+        return isBookTitleNull() || book.getTitle().toLowerCase().contains(getBookTitle());
+    }
+}
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java:44]: please add a newline

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        bookGrid.setItems(filterShelf(matchingShelves.get(0), bookFilters));
+    }
+
+    private List<Book> filterShelf(PredefinedShelf selectedShelf, BookFilters bookFilters) {
+        return selectedShelf.getBooks()
+                            .stream()
+                            .filter(book -> bookFilters.apply(book))
+                            .collect(Collectors.toList());
+    }
+}
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java:81]: another newline to be added

File: src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java [hunk 10]
```diff
@@ -35,11 +26,7 @@
 import com.vaadin.flow.component.UI;
 import com.vaadin.flow.component.grid.Grid;
 import com.vaadin.flow.spring.SpringServlet;
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.Assertions;
-import org.junit.jupiter.api.Assumptions;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java:29]: Please undo the wildcard import change

File: src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java [hunk 11]
```diff
@@ -51,13 +38,7 @@
...
 import java.util.List;
 
-import static com.karankumar.bookproject.ui.shelf.BooksInShelfView.AUTHOR_KEY;
-import static com.karankumar.bookproject.ui.shelf.BooksInShelfView.DATE_FINISHED_KEY;
-import static com.karankumar.bookproject.ui.shelf.BooksInShelfView.DATE_STARTED_KEY;
-import static com.karankumar.bookproject.ui.shelf.BooksInShelfView.GENRE_KEY;
-import static com.karankumar.bookproject.ui.shelf.BooksInShelfView.PAGES_KEY;
-import static com.karankumar.bookproject.ui.shelf.BooksInShelfView.RATING_KEY;
-import static com.karankumar.bookproject.ui.shelf.BooksInShelfView.TITLE_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java:41]: Please undo the wildcard import change

File: src/test/java/com/karankumar/bookproject/ui/goal/ReadingGoalViewTests.java [hunk 12]
```diff
@@ -7,23 +7,15 @@
...
 import com.vaadin.flow.component.UI;
 import com.vaadin.flow.spring.SpringServlet;
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.Assertions;
-import org.junit.jupiter.api.Assumptions;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/goal/ReadingGoalViewTests.java:16]: Please undo the wildcard import change

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java [hunk 13]
```diff
@@ -17,23 +17,14 @@
...
 import com.karankumar.bookproject.ui.MockSpringServlet;
 import com.vaadin.flow.component.UI;
 import com.vaadin.flow.spring.SpringServlet;
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.Assertions;
-import org.junit.jupiter.api.Assumptions;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java:27]: Please undo the wildcard import change

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java [hunk 14]
```diff
@@ -17,23 +17,14 @@
 
 import com.github.mvysny.kaributesting.v10.MockVaadin;
 import com.github.mvysny.kaributesting.v10.Routes;
-import com.karankumar.bookproject.backend.entity.Author;
-import com.karankumar.bookproject.backend.entity.Book;
-import com.karankumar.bookproject.backend.entity.Genre;
-import com.karankumar.bookproject.backend.entity.PredefinedShelf;
-import com.karankumar.bookproject.backend.entity.RatingScale;
+import com.karankumar.bookproject.backend.entity.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java:20]: Please undo the wildcard import change

**Comment by kaansonmezoz** [comment]
I can divide the refactorings into multiple issues & open PR's to them  If it's okay with you. I could just cherry-pick commits into related issue. And make this PR simpler. 

What do you say ? :)

**Comment by knjk04** [comment]
My bad. I was thinking you wanted to refactor the whole codebase in this single PR :) (since there were a lot of unrelated changes -- refactoring in the entities, repositories, services, security classes, etc.)

Let's keep this PR and work on it. If you could please start by undoing the wildcard import changes as well any changes where the ordering of imports was changed, it should hopefully reduce the number of classes to review.

Lastly, if I made you feel bad (I noticed the sad emoji) with my review, I apologise. That was not my intention. I was just trying to help make this PR better and ensure the project remains clean :)

**Comment by kaansonmezoz** [comment]
Oh no man, I was upset because my PR got bigger and hard to read. It wasn't about you. Your feedbacks are important 🙂

**Comment by knjk04** [comment]
Before continuing with this, I recommend first resolving the merge conflicts and then rebasing. 

I've done a lot of refactoring, so you may even want to start afresh.

**Comment by kaansonmezoz** [comment]
I've been busy recently so I couldn't find a chance to resolve issues. I'm starting with resolving the conflicts between master

File: src/main/java/com/karankumar/bookproject/backend/entity/Author.java (hunk 1 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/entity/Author.java]: Done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/backend/entity/Role.java (hunk 2 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/entity/Role.java]: Done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/backend/entity/Shelf.java (hunk 3 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/entity/Shelf.java]: Done :slightly_smiling_face:

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java (hunk 14 above)
Comment (kaansonmezoz) [reply src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java]: done :slightly_smiling_face:

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java (hunk 13 above)
Comment (kaansonmezoz) [reply src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java]: done :slightly_smiling_face:

File: src/test/java/com/karankumar/bookproject/ui/goal/ReadingGoalViewTests.java (hunk 12 above)
Comment (kaansonmezoz) [reply src/test/java/com/karankumar/bookproject/ui/goal/ReadingGoalViewTests.java]: done :slightly_frowning_face:

File: src/main/java/com/karankumar/bookproject/backend/entity/User.java (hunk 4 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/entity/User.java]: done :slightly_frowning_face:

File: src/main/java/com/karankumar/bookproject/backend/entity/User.java (hunk 5 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/entity/User.java]: done :slightly_frowning_face:

File: src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java (hunk 10 above)
Comment (kaansonmezoz) [reply src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java]: done :slightly_smiling_face:

File: src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java (hunk 11 above)
Comment (kaansonmezoz) [reply src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTests.java]: done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java (hunk 8 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java]: done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java (hunk 9 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java]: done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/ui/components/PaperToggle.java (hunk 7 above)
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/components/PaperToggle.java]: This one still needs to be undone

File: src/main/java/com/karankumar/bookproject/ui/shelf/component/BookGridColumn.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+    private void addDateFinishedColumn() {
+        bookGrid.addColumn(createLocalDateRenderer(Book::getDateFinishedReading))
+                .setHeader("Date finished reading")
+                .setComparator(Comparator.comparing(Book::getDateStartedReading))
+                .setSortable(true)
+                .setKey(DATE_FINISHED_KEY);
+    }
+
+    private LocalDateRenderer createLocalDateRenderer(ValueProvider<Book, LocalDate> provider) {
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/shelf/component/BookGridColumn.java:85]: Instead of using the raw type, can you generify this to use `Book`?

File: src/main/java/com/karankumar/bookproject/ui/shelf/component/BookShelfComboBox.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
+package com.karankumar.bookproject.ui.shelf.component;
+
+import com.karankumar.bookproject.backend.entity.PredefinedShelf;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/shelf/component/BookShelfComboBox.java:3]: remove this unused import

File: src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTest.java [hunk 17]
```diff
@@ -43,6 +43,15 @@
...
 import java.util.List;
 
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.AUTHOR_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.DATE_FINISHED_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.DATE_STARTED_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.GENRE_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.PAGES_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.RATING_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.TITLE_KEY;
+import static com.karankumar.bookproject.ui.shelf.component.BookGridColumn.PAGES_READ_KEY;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTest.java:53]: We're not using static imports

File: src/main/java/com/karankumar/bookproject/backend/entity/Book.java [hunk 18]
```diff
@@ -108,4 +108,8 @@ public String toString() {
             + "title='" + title + '\''
             + '}';
     }
+
+    public boolean existsSeriesPosition() {
```
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java:112]: done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/backend/repository/TagRepository.java [hunk 19]
```diff
@@ -1,10 +1,9 @@
 package com.karankumar.bookproject.backend.repository;
 
+import com.karankumar.bookproject.backend.entity.Tag;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
-import com.karankumar.bookproject.backend.entity.Tag;
```
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/repository/TagRepository.java:6]: done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.karankumar.bookproject.backend.entity.Book;
+
+class BookFilters {
+    private String bookTitle;
+    private String bookAuthor;
+
+    private String getBookTitle() {
+        return bookTitle.toLowerCase();
+    }
```
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java:11]: I believe it will be less readable if  we inline it. I think it's more import to have a readable code rather than smaller code base :slightly_smiling_face:  But sure let's do it :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/backend/service/TagService.java [hunk 21]
```diff
@@ -1,14 +1,12 @@
...
-import org.springframework.stereotype.Service;
-
 import com.karankumar.bookproject.backend.entity.Tag;
 import com.karankumar.bookproject.backend.repository.TagRepository;
-
 import lombok.extern.java.Log;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.logging.Level;
```
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/service/TagService.java:9]: done :slightly_smiling_face: :

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java]: As a compromise, what about changing the method name to something like `getCaseInsensitiveBookTitle()`?

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 22]
```diff
@@ -57,307 +55,129 @@
...
-        filterByAuthorNameField.setPlaceholder("Filter by Author Name");
-        filterByAuthorNameField.setClearButtonVisible(true);
-        filterByAuthorNameField.setValueChangeMode(ValueChangeMode.LAZY);
-        filterByAuthorNameField.addValueChangeListener(eventFilterAuthorName -> {
-            if (eventFilterAuthorName.getValue() != null) {
-                authorNameToFilterBy = eventFilterAuthorName.getValue();
-            }
-            updateGrid();
-        });
+        visibilityStrategies.get(predefinedShelfName).toggleColumnVisibility(bookGrid);
```
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java:160]: done :+1:

File: src/main/java/com/karankumar/bookproject/backend/entity/Tag.java [hunk 23]
```diff
@@ -1,18 +1,17 @@
...
 import javax.persistence.FetchType;
 import javax.persistence.ManyToMany;
 import javax.validation.constraints.NotEmpty;
 import javax.validation.constraints.NotNull;
-
-import lombok.AccessLevel;
-import lombok.Data;
-import lombok.EqualsAndHashCode;
-import lombok.NoArgsConstructor;
+import java.util.Set;
```
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/backend/entity/Tag.java]: done :+1:

File: src/main/java/com/karankumar/bookproject/ui/shelf/component/BookGridColumn.java (hunk 15 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/ui/shelf/component/BookGridColumn.java]: done :+1:

File: src/main/java/com/karankumar/bookproject/ui/shelf/component/BookShelfComboBox.java (hunk 16 above)
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/ui/shelf/component/BookShelfComboBox.java]: done :slightly_smiling_face:

File: src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTest.java (hunk 17 above)
Comment (kaansonmezoz) [reply src/test/java/com/karankumar/bookproject/ui/shelf/BooksInShelfViewTest.java]: done :slightly_smiling_face:

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookFilters.java]: I'm glad that you tried to meet halfway :slightly_smiling_face:  But I changed my mind and decided to go with your way :slightly_smiling_face:

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 24]
```diff
@@ -13,8 +13,9 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import java.net.*;
+
 import java.io.*;
+import java.net.*;
```
[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply .mvn/wrapper/MavenWrapperDownloader.java:18]: You are absolutely right. I used "Optimze Import" tool of IntelliJ Idea. I think it changed imports :slightly_frowning_face:

[the remark this answers is not in the document]
Comment (kaansonmezoz) [reply .mvn/wrapper/MavenWrapperDownloader.java]: done :slightly_smiling_face:

**Comment by knjk04** [comment]
Let me know when you want another review

**Comment by kaansonmezoz** [comment]
My PR is ready to be reviewed again. I've changed some of the commit messages. But I didn't want to change older commit messages, since rebasing them will cause some commits to be duplicated :slightly_frowning_face:  Since I merged from master to resolve conflicts :slightly_frowning_face: But I can still override those messages via rebasing if it's okay to duplicate some of those commits

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 25]
```diff
@@ -57,307 +55,129 @@
...
     }
 
-    private void saveCustomShelf(CustomShelfForm.SaveEvent event) {
-        if (event.getCustomShelf() != null) {
-            customShelfService.save(event.getCustomShelf());
-            LOGGER.log(Level.INFO, "Custom shelf saved");
-        } else {
-            LOGGER.log(Level.SEVERE, "Custom shelf value is null");
-        }
+    List<Grid.Column<Book>> getColumns() {
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java:182]: I've added this just now

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 26]
```diff
@@ -57,307 +55,129 @@
...
+        CustomShelfUtils shelfUtils = new CustomShelfUtils(customShelfService);
+        bookGrid.update(chosenShelf, shelfUtils, bookFilters, predefinedShelfUtils);
     }
 
-    private void editBook(Book book) {
-        if (book != null && bookForm != null) {
-            bookForm.setBook(book);
-            bookForm.openForm();
-        }
+    public void chooseShelf(String chosenShelf) {
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java]: Added

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 27]
```diff
@@ -57,307 +55,129 @@
...
-    private void resetGridColumns() {
-        bookGrid.setColumns();
+        return layout;
     }
 
-    private void updateGrid() {
-        if (chosenShelf == null) {
-            LOGGER.log(Level.FINEST, "Chosen shelf is null");
+    /**
+     * @throws NotSupportedException if a shelf is not supported.
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java]: Removed

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 28]
```diff
@@ -57,307 +55,129 @@
...
-        addPagesReadColumn();
+    private void bindListeners(BookService bookService) {
+        new BookSaveListener(bookService, this).bind(bookForm);
+        new BookDeleteListener(bookService, this).bind(bookForm);
     }
 
-    private void addPagesColumn() {
-        bookGrid.addColumn(BookGridKeys.PAGES_KEY);
-    }
+    private EnumMap<PredefinedShelf.ShelfName, BookVisibilityStrategy> initVisibilityStrategies() {
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java:102]: Moved

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 29]
```diff
@@ -57,307 +55,129 @@
...
+    private final BookFilters bookFilters;
 
     private final PredefinedShelfUtils predefinedShelfUtils;
 
-    public BooksInShelfView(BookService bookService, PredefinedShelfService predefinedShelfService,
+    public BooksInShelfView(BookService bookService, PredefinedShelfService shelfService,
                             CustomShelfService customShelfService) {
-        this.bookService = bookService;
-        this.predefinedShelfService = predefinedShelfService;
+
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java:76]: Deleted

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    private void populateGridWithBooks(Set<Book> books, BookFilters bookFilters) {
+        List<Book> items = filterShelf(books, bookFilters);
+        bookGrid.setItems(items);
+    }
+
+    private List<Book> filterShelf(Set<Book> books, BookFilters bookFilters) {
+        return books.stream()
+                .filter(book -> bookFilters.apply(book))
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java]: I've madet these changed

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+            if (book != null && bookForm != null) {
+                bookForm.setBook(book);
+                bookForm.openForm();
+            }
+        });
+    }
+
+    public void toggleColumnVisibility(String columnKey, boolean showColumn) {
+        if (bookGrid.getColumnByKey(columnKey) == null) {
+            LOGGER.log(Level.SEVERE, "Key is null:" + columnKey);
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java]: Added space

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+        } else {
+            bookGrid.getColumnByKey(columnKey).setVisible(showColumn);
+        }
+    }
+
+    public Grid<Book> get() {
+        return bookGrid;
+    }
+
+    void update(String chosenShelf, CustomShelfUtils customShelfUtils, BookFilters bookFilters, PredefinedShelfUtils predefinedShelfUtils) {
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java]: Moved to constructor

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+                bookForm.openForm();
+            }
+        });
+    }
+
+    public void toggleColumnVisibility(String columnKey, boolean showColumn) {
+        if (bookGrid.getColumnByKey(columnKey) == null) {
+            LOGGER.log(Level.SEVERE, "Key is null:" + columnKey);
+        } else {
+            bookGrid.getColumnByKey(columnKey).setVisible(showColumn);
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java:50]: Done

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 34]
```diff
@@ -21,33 +21,31 @@
...
-import com.karankumar.bookproject.backend.utils.ShelfUtils;
 import com.karankumar.bookproject.ui.MainView;
 import com.karankumar.bookproject.ui.book.BookForm;
+import com.karankumar.bookproject.ui.shelf.component.AuthorFilterText;
+import com.karankumar.bookproject.ui.shelf.component.BookShelfComboBox;
+import com.karankumar.bookproject.ui.shelf.component.TitleFilterText;
+import com.karankumar.bookproject.ui.shelf.listener.BookDeleteListener;
+import com.karankumar.bookproject.ui.shelf.listener.BookSaveListener;
+import com.karankumar.bookproject.ui.shelf.listener.CustomShelfListener;
+import com.karankumar.bookproject.ui.shelf.visibility.*;
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java]: Removed

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 35]
```diff
@@ -21,33 +21,31 @@
...
-import java.time.format.FormatStyle;
-import java.util.Comparator;
-import java.util.Set;
-import java.util.function.Predicate;
-import java.util.logging.Level;
-import java.util.stream.Collectors;
+import java.util.EnumMap;
+import java.util.List;
+
+import static com.karankumar.bookproject.backend.entity.PredefinedShelf.ShelfName.*;
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java]: Removed

File: src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java [hunk 36]
```diff
@@ -57,307 +55,129 @@
...
+    public final BookGrid bookGrid;
+    public final BookShelfComboBox whichShelf;
 
-    private final BookForm bookForm;
+    private final EnumMap<PredefinedShelf.ShelfName, BookVisibilityStrategy> visibilityStrategies;
 
-    private final BookService bookService;
-    private final PredefinedShelfService predefinedShelfService;
+    private final BookForm bookForm;
+    private final PredefinedShelfService shelfService;
```
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/shelf/BooksInShelfView.java]: Done

**Comment by knjk04** [comment]
> 
> 
> My PR is ready to be reviewed again. I've changed some of the commit messages. But I didn't want to change older commit messages, since rebasing them will cause some commits to be duplicated 🙁 Since I merged from master to resolve conflicts 🙁 But I can still override those messages via rebasing if it's okay to duplicate some of those commits

Don't worry about it then. Just something to bear in mind for next time :)

**Comment by knjk04** [comment]
For future reference, a more concise branch name would have been: `refactor/bookshelf-view-197`. In your current branch name, you have the word `issue`, which is redundant, and you mention refactoring twice. 

I've tried renaming the branch, but I don't have permission to push to your fork. I'm just going to merge this in now as someone else is waiting on this to be completed, but it's just something to note for next time.

Thanks again for all of your help!
